### PR TITLE
Fix documentation

### PR DESCRIPTION
**Pull request by maukoquiroga**
Partially openfisca/openfisca-doc#244
Depended upon by #1033 #1021 

#### New Features

- Introduce `openfisca_core.variables.typing`
  - Documents the signature of formulas
  - Note: as formulas are generated dynamically, documenting them is tricky

#### Bug Fixes

- Fix the official doc
  - Corrects malformed docstrings
  - Fix missing and/ou outdated references

#### Technical Changes

- Add tasks to automatically validate that changes do not break the documentation

#### Documentation

- Add steps to follow in case the documentation is broken
- Add general documenting guidelines in CONTRIBUTING.md

Capture d’écran 2021-08-26 à 16 57 13

### CODE REVIEW

**Comment by benjello** [comment]
@maukoquiroga got this error when running `make doc.test`:

```
Warning, treated as error:
/home/mbenjelloul/openfisca/openfisca-core/openfisca_core/indexed_enums/enum.py:docstring of openfisca_core.indexed_enums.Enum.encode:None: WARNING: py:obj reference target not found: numpy.ndarray
```

**Comment by maukoquiroga** [comment]
Hi @benjello ! As I put in the README.md, you have to run:

```
make test.doc.build branch=`git branch --show-current`
```

So it compares these changes with the corresponding ones in the doc repo.

Please let me know if that works for you, and if and how could I render the documentation clearer.

**Comment by maukoquiroga** [comment]
By the way @MattiSG I did actually struggle to get this automated with `make`, and I have been practicing stoicism so I am in no (hard) way emotionally attached to keep this overly complex and fragile task, quite the contrary, I just hope it won't be a blocking factor for this specific changeset 😃

File: README.md [hunk 1]
```diff
@@ -96,6 +96,76 @@ exec make format-style
...
+Yet however OpenFisca does not follow a common convention for docstrings, our current toolchain allows to check whether documentation builds correctly and to update it automatically with each contribution to this repository.
+
+In the meantime, please take a look at our [contributing guidelines](CONTRIBUTING.md) for some general tips on how to document your contributions, and at our official documentation's [repository](https://github.com/openfisca/openfisca-doc/blob/master/README.md) to in case you want to know how to build it by yourself —and improve it!
+
+### To verify that the documentation still builds correctly
+
+You can run:
+
+```sh
+make test.doc
```
Comment (benjello) [inline README.md:110]: @maukoquiroga : I actually did follow the README and got stuck in here ;-)

Comment (maukoquiroga) [reply README.md:110]: Thanks for the feedback, I've pushed a fix, it should work now :)

**Review by HAEKADI — COMMENTED** [review]
Thank you @maukoquiroga :) I tested the `make` commands and they're functional.
I don't know if I can approve this PR though, simply because a lot of work has been done in the `makefile` and I don't speak `make` very well, so I can't be of much help 😅

**Review by benjello — DISMISSED** [review]

**Comment by benjello** [comment]
Thanks @maukoquiroga ! I tested make test-doc and it worked well.
I didn't test the occurrence of an error and how to fix it, but I am all in afvor of merging this PR and improving it if needed later when use case emerge.

**Review by benjello — APPROVED** [review]

**Comment by benjello** [comment]
@HAEKADI, I assume that  you are almost aligned with me ;-) (or do not have a strong opinion which is equivalent)
Review from @MattiSG and @sandcha might be useful as well ! ()

**Pull request merged**